core_lib/src/interface/timelinecells.cpp

```diff
@@ -860,6 +860,11 @@ void TimeLineCells::mouseReleaseEvent(QMouseEvent* event)
     {
         Layer *currentLayer = mEditor->object()->getLayer(layerNumber);
 
+        if (mMovingFrames) {
+            emit framesMoved();
+
+            mMovingFrames = false;
+        }
```

Comment:
I don't really like where this is being called from. We're invalidating the cache because of changes in the keyframe position, not directly because of the user's interaction with the timeline. A signal from editor or layer would be better I think.

If we edit Layer::moveSelectedFrames instead, we can call updateFrame for each frame affected (all frames between the old frame position and the new frame position inclusive). This might add a bit of overhead to the dragging interaction, but I think it's worth it for preserving the rest of the frames, and that will be a moot point if/when we merge #1495 since that will only have to be done once after the frames have been dropped.

Reply:
We could do that but is it necessary? You're not looking at those frames at the given time you're moving, you're only looking at one frame and paint x frames behind and in-front with onion skin enabled, so it should only be necessary to update the current frame and run through the frames affected by onion skin once.

The problem I have with calling update for all affected frames is that we either have to make a new updateFrame function or modify it, as it will otherwise call update onionSkinAround for every affected frame. This should only be necessary for the frame you're looking at, yet we'll be doing it multiple times.

it's easy to clear the cache here and I don't see a big reason to go through hoops to affect specified frames however I'll look into another way of updating, maybe introduce a new cache related flag, to make sure we're not clearing more than needed.

I don't think Layer should have an emitter though, it's just a model, this should exist elsewhere, I agree through that Timeline may not have been the best place, even though it's the closest to the user interface and therefore easiest change without causing side-effects.

Reply:
> You're not looking at those frames at the given time you're moving, you're only looking at one frame and paint x frames behind and in-front with onion skin enabled, so it should only be necessary to update the current frame and run through the frames affected by onion skin once.

No it is not necessary to update the view any frame other than the current one while dragging. The point of what I'm suggesting is not to call it while dragging because they need to be updated before the frame is dropped, but rather because it is the simplest way to implement what you suggest but have not done: update only the affected frames. You can't do this easily in mouseReleaseEvent because you don't know where the frame originated from. You could add logic to handle this and then iterate through the frames inbetween the start and end positions, but why would you when a lot of that could be handled by the changes in #1495?

> as it will otherwise call update onionSkinAround for every affected frame. This should only be necessary for the frame you're looking at, yet we'll be doing it multiple times.

No, this is not enough. At the very least updateOnionSkinsAround needs to be called for the initial and final positions of all selected frames.

> it's easy to clear the cache here and I don't see a big reason to go through hoops to affect specified frames however I'll look into another way of updating, maybe introduce a new cache related flag, to make sure we're not clearing more than needed.

The reason to jump through hoops here is that rendering even a single frame is likely far more expensive than the most inefficient methods of finding the frames which have been modified by a move. I haven't benchmarked this, but intuitively even a few thousand checks in the cache would be nothing compared to the millions of pixels that need to be potentially retrieved from a disk, decoded, allocated, copied, composited, etc. for each and every frame we invalidate and subsequently have to re-render. My philosophy is that we should be using the caches as much as possible, even at the cost of additional code complexity. Playback is arguably our most important functionality, and we need to do more than what we have been doing to assure realtime playback (especially for higher frame rates or lower-powered systems). I'll admit the initial implementation of it has had many problems, but I think we can get this right and I think it will be worth it.

> I don't think Layer should have an emitter though, it's just a model, this should exist elsewhere, I agree through that Timeline may not have been the best place.

Yes you're right, Layer might not be the best choice. I think you understand my issue with where it is currently though. This also ties into why did not suggest doing a single pass at the end of the dragging operation: only the timeline knows if a frame is actively being dragged so I'm not sure how you would do that from anywhere else.